tensorpipe/test/channel/channel_test.cc

```diff
@@ -15,495 +15,556 @@
-        auto descriptor = peers_->recv(PeerGroup::kClient);
-        std::future<Error> recvFuture =
-            recvWithFuture(channel, descriptor, CpuBuffer{nullptr, 0});
-        Error recvError = recvFuture.get();
-        EXPECT_FALSE(recvError) << recvError.what();
-
-        peers_->done(PeerGroup::kClient);
-        peers_->join(PeerGroup::kClient);
-
-        ctx->join();
-      });
-}
-
-TEST_P(ChannelTest, EmptyTensor) {
-  // Call send and recv with a length of 0 but a non-null pointer.
-
-  testConnection(
-      [&](std::shared_ptr<transport::Connection> conn) {
-        std::shared_ptr<CpuContext> ctx = GetParam()->makeContext("server");
-        auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
-
-        // Allocate a non-empty vector so that its .data() pointer is non-null.
-        std::vector<uint8_t> data(1);
-
-        // Perform send and wait for completion.
-        std::future<std::tuple<Error, TDescriptor>> descriptorFuture;
-        std::future<Error> sendFuture;
-        std::tie(descriptorFuture, sendFuture) =
-            sendWithFuture(channel, CpuBuffer{data.data(), 0});
-        Error descriptorError;
-        TDescriptor descriptor;
-        std::tie(descriptorError, descriptor) = descriptorFuture.get();
-        EXPECT_FALSE(descriptorError) << descriptorError.what();
-        peers_->send(PeerGroup::kClient, descriptor);
-        Error sendError = sendFuture.get();
-        EXPECT_FALSE(sendError) << sendError.what();
-
-        peers_->done(PeerGroup::kServer);
-        peers_->join(PeerGroup::kServer);
-
-        ctx->join();
-      },
-      [&](std::shared_ptr<transport::Connection> conn) {
-        std::shared_ptr<CpuContext> ctx = GetParam()->makeContext("client");
-        auto channel = ctx->createChannel(std::move(conn), Endpoint::kConnect);
-
-        // Allocate a non-empty vector so that its .data() pointer is non-null.
-        std::vector<uint8_t> data(1);
-
-        // Perform recv and wait for completion.
-        auto descriptor = peers_->recv(PeerGroup::kClient);
-        std::future<Error> recvFuture =
-            recvWithFuture(channel, descriptor, CpuBuffer{data.data(), 0});
-        Error recvError = recvFuture.get();
-        EXPECT_FALSE(recvError) << recvError.what();
-
-        peers_->done(PeerGroup::kClient);
-        peers_->join(PeerGroup::kClient);
-
-        ctx->join();
-      });
-}
-
-TEST_P(ChannelTest, contextIsNotJoined) {
+  void server(std::shared_ptr<transport::Connection> conn) override {
+    std::shared_ptr<Context<TBuffer>> ctx =
+        this->helper_->makeContext("server");
+    auto channel = ctx->createChannel(std::move(conn), Endpoint::kListen);
+
+    // Perform send and wait for completion.
+    std::future<std::tuple<Error, TDescriptor>> descriptorFuture;
+    std::future<Error> sendFuture;
+    std::tie(descriptorFuture, sendFuture) =
+        sendWithFuture(channel, TBuffer{nullptr, 0});
```

Comment:
This is not a "generic" test, because it assumes that TBuffer can be constructed using two arguments: a pointer and a length. This may be the case with Cpu and CudaBuffer, but it's not guaranteed to always be the case. This test aims to test whether a channel can handle a null pointer but, in principle, a TBuffer may not even hold a pointer. So maybe it's better to make this become a Cpu-specific test? (And possibly create a copy of it for Cuda channels, if we decide we want Cuda channels to also be able to handle null pointers)